### PR TITLE
Better status condition management

### LINKED ISSUE #41: Better status condition management
With https://github.com/strimzi/kafka-access-operator/pull/37,  the  KafkaAccess resource has now a `Ready` condition, however , the `setLastTransitionTime` is always updated with the invocation time of the [setReadyCondition](https://github.com/strimzi/kafka-access-operator/blob/50a2489e0f2f0685e649bcc756847f6f346a60bd/src/main/java/io/strimzi/kafka/access/model/KafkaAccessStatus.java#L68) method, however such field is meant to report the last time the condition changed from one status to another. 

A side effect is that, since the status always changes, then additional events are generated by the API Server and observing controllers may have to deal with superflous updates.

Ideally we could follow what it is done in the Kubernetes [apimachinery codebase ](https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/conditions.go#L25-L68).

I can work on a PR  if this make sense.

### PR DESCRIPTION
**Pull request by lburgazzoli**
Fixes #41

### CODE REVIEW

File: src/test/java/io/strimzi/kafka/access/model/StatusUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            .withType(StatusUtils.CONDITION_TYPE_READY)
+            .withStatus(StatusUtils.CONDITION_STATUS_TRUE)
+            .withMessage("foo-4")
+            .withReason("bar")
+            .build();
+
+        List<Condition> conditions = new ArrayList<>();
+        conditions.add(c1);
+
+        {
```
Comment (katheris) [inline src/test/java/io/strimzi/kafka/access/model/StatusUtilsTest.java:47]: I'd prefer if these test cases were broken out into separate, smaller test cases with an `@DisplayName` annotation explaining the aim of the test. That way they will be easier to refactor in future if the StatusUtils class is updated and if something breaks the test it's quicker to see exactly which behaviour is broken

**Review by katheris — COMMENTED** [review]
Hi @lburgazzoli thanks for your PR, the implementation looks good, I just had a few comments about the package for StatusUtils and the test structure.

File: src/main/java/io/strimzi/kafka/access/model/StatusUtils.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.kafka.access.model;
```
[the remark this answers is not in the document]
Comment (lburgazzoli) [reply src/main/java/io/strimzi/kafka/access/model/StatusUtils.java]: done

File: src/test/java/io/strimzi/kafka/access/model/StatusUtilsTest.java (hunk 1 above)
Comment (lburgazzoli) [reply src/test/java/io/strimzi/kafka/access/model/StatusUtilsTest.java]: done.

let me know if it matches the expectation

File: src/test/java/io/strimzi/kafka/access/internal/StatusUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        StatusUtils.setCondition(conditions, c2);
+
+        assertThat(conditions).first().satisfies(c -> {
+            assertThat(c.getLastTransitionTime()).isNotNull();
+            assertThat(c.getMessage()).isEqualTo(c2.getMessage());
+        });
+    }
+
+    @Test
+    @DisplayName("When a condition is amended and it status changes, then the LastTransitionTime should be updated according")
```
Comment (katheris) [inline src/test/java/io/strimzi/kafka/access/internal/StatusUtilsTest.java:66]:
```suggestion
    @DisplayName("When a condition is amended and its status changes, then the LastTransitionTime should be updated according")
```

File: src/test/java/io/strimzi/kafka/access/internal/StatusUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+        assertThat(conditions).first().satisfies(c -> {
+            assertThat(c.getLastTransitionTime()).isNotNull();
+            assertThat(c.getLastTransitionTime()).isNotSameAs(ltt1);
+            assertThat(c.getMessage()).isEqualTo(c2.getMessage());
+        });
+    }
+
+    @Test
+    @DisplayName("When a condition is amended and it status does not change, then the LastTransitionTime should not be ièdated")
```
Comment (katheris) [inline src/test/java/io/strimzi/kafka/access/internal/StatusUtilsTest.java:103]:
```suggestion
    @DisplayName("When a condition is amended and its status does not change, then the LastTransitionTime should not be updated")
```

File: src/test/java/io/strimzi/kafka/access/internal/StatusUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        // add a condition, LastTransitionTime should be set
+        StatusUtils.setCondition(conditions, c1);
+
+        assertThat(conditions).first().satisfies(c -> {
+            assertThat(c.getLastTransitionTime()).isNotNull();
+        });
+
+        var ltt1 = conditions.get(0).getLastTransitionTime();
+
+        // amend the condition but not ots status, LastTransitionTime should not be updated
```
Comment (katheris) [inline src/test/java/io/strimzi/kafka/access/internal/StatusUtilsTest.java:129]:
```suggestion
        // amend the condition but not its status, LastTransitionTime should not be updated
```

**Review by katheris — COMMENTED** [review]
Thanks @lburgazzoli those tests look better, couple of small nits, but otherwise the PR looks good to me

File: src/test/java/io/strimzi/kafka/access/internal/StatusUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+import io.strimzi.api.kafka.model.status.Condition;
+import io.strimzi.api.kafka.model.status.ConditionBuilder;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class StatusUtilsTest {
+    @Test
+    @DisplayName("When a condition is added, then the LastTransitionTime should be set if empty")
```
Comment (PaulRMellor) [inline src/test/java/io/strimzi/kafka/access/internal/StatusUtilsTest.java:19]:
```suggestion
    @DisplayName("When a condition is added, the LastTransitionTime should be set if empty")
```

File: src/test/java/io/strimzi/kafka/access/internal/StatusUtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+        StatusUtils.setCondition(conditions, c1);
+
+        assertThat(conditions).first().satisfies(c -> {
+            assertThat(c.getLastTransitionTime()).isNotNull();
+        });
+    }
+
+    @Test
+    @DisplayName("When a condition is amended, then the LastTransitionTime should be set if empty")
```
Comment (PaulRMellor) [inline src/test/java/io/strimzi/kafka/access/internal/StatusUtilsTest.java:38]:
```suggestion
    @DisplayName("When a condition is amended, the LastTransitionTime should be set if empty")
```

File: src/test/java/io/strimzi/kafka/access/internal/StatusUtilsTest.java (hunk 3 above)
Comment (PaulRMellor) [inline src/test/java/io/strimzi/kafka/access/internal/StatusUtilsTest.java:66]:
```suggestion
    @DisplayName("When a condition is amended and its status changes, the LastTransitionTime should be updated accordingly")
```

File: src/test/java/io/strimzi/kafka/access/internal/StatusUtilsTest.java (hunk 4 above)
Comment (PaulRMellor) [inline src/test/java/io/strimzi/kafka/access/internal/StatusUtilsTest.java:103]:
```suggestion
    @DisplayName("When a condition is amended and its status does not change, the LastTransitionTime should not be updated")
```

**Review by PaulRMellor — APPROVED** [review]
Looks good. I made a couple of suggestions for the descriptions.

**Comment by katheris** [comment]
@lburgazzoli can you make the tweaks Paul and I suggested? Then we can look to get this PR merged

**Comment by lburgazzoli** [comment]
> @lburgazzoli can you make the tweaks Paul and I suggested? Then we can look to get this PR merged

@katheris done (sorry for the delay, just got back from PTO)

**Review by katheris — APPROVED** [review]

**Pull request merged**